src/sbd-inquisitor.c

```diff
@@ -41,6 +41,10 @@ void recruit_servant(const char *devname, pid_t pid)
 	struct servants_list_item *s = servants_leader;
 	struct servants_list_item *newbie;
 
+	if (!devname) {
+	    cl_log(LOG_ERR, "Servant needs a name to be recruited");
+	    return;
+	}
```

Comment:
Redundant in the current state.

Reply:
Btw. this is the problem of not binding to some sort of documented
contract, like

```
device name comes from user input (configuration, command-line switch)
or from built-in literal, and all these cases are ensured to be no-null
before passing them further around sbd logic
```

Such unified-assumptions dealing would prevent such paranoia (and
arising noise in code).  Asserts can be still used to as a last-resort
checker and reminder of such contracts.

Reply:
Well - more comprehensive logging and you can use devname in the logs further down as you've just discovered ;-)

Reply:
It's redundant because the unsaid contract already makes it like
that, so retesting won't bring anything new unless you break that
contract.  You can keep retesting it all the way down, at the expense
of horrible code.

Reply:
Well, when those contracts are not written down or not implemented properly - as here - you are right the NULL is in fact prevented by using it in sscanf causing some undefined behavior rather then the following NULL check ...
Anyway you have to stop it at some point if you don't want to do a complete overhaul in one go and as you said we wanted something quick ;-)

Reply:
and you would probably want a consistent reaction on heap allocation failing ... something dramatic as exiting out isn't bad as done in some places ...

Reply:
maybe some kind of assert-macro after a global overhaul on this issue.
caveat is that bailing out is fine during config but upon watcher-operation we have to trigger the timeout_action.

Reply:
That converges towards "who's supervising the supervisor" meta-discussion
that is nonetheless important -- what if sbd crashes? same situation.
I have no idea about the failure modes of sbd (and that's, not very
conveniently, quite an uncharted territory with pacemaker, too), original
`assert` proposal was silently betting on watchdog to save the situation.

Reply:
BTW. as I've mentioned with
https://github.com/ClusterLabs/pacemaker/pull/1681/commits/a2e873635db3dfbb696527372dfaad9f58621f48,
there are nonnull annotations to help guarding such contracts
as well.